### ISSUE TITLE
formControlName: No value accessor for form control with path

### ISSUE THREAD

**Issue by AppiePau** [KillerCodeMonkey/ngx-quill#1212]
Hi,

I'm trying to get ngx-quill working in my project for the first time, using Angular 11. It all seems to look good, but binding using formControlName seems to fail.  The following (simplified) HTML code has replaced my textarea code that works fine with the formControlName attribute :

```html
<form *ngIf="form" [formGroup]="form" (submit)="submit()">
    <div formArrayName="templates">
        <div class="card mt-3" *ngFor="let template of templates.controls; let i=index;"
            [formGroupName]="i">
            ...
            <div class="card-body">
                ....
                    <quill-editor #editor [modules]="modules" 
                            (onEditorCreated)="editorCreated($event)" placeholder="Enter Text"
                            (onSelectionChanged)="onSelectionChanged($event)"
                            (onContentChanged)="onContentChanged($event)" name="body"
                            formControlName="body"></quill-editor>
                ....
            </div>
        </div>
    </div>
</form>
```
The full error I get is:
```
core.js:6157 ERROR Error: No value accessor for form control with path: 'templates -> 0 -> body'
```

I know that error from past experiences building my own controls. I noticed the following fixed the issue with my controls, but is not implemented in ngx-quill:

```ts
  constructor(private control: NgControl) {
    this.control.valueAccessor = this;
  }
```

Could you let me know what I'm doing wrong or perhaps my suggestion helps to solve the problem? Let me know if you need some more info.

Edit, these are my used packages:
```
    "ngx-quill": "^13.2.0",
    "quill": "^1.3.7",
```

Cheers,
Paul

**Comment by KillerCodeMonkey** [comment]
Are you Sure your Control exists on form[0] ?

Here you have a working example without The Formarray

And it IS working
https://github.com/KillerCodeMonkey/ngx-quill-example/blob/master/src/app/reactive-forms/reactive-forms.component.html#L19


Am 11.04.2021 in 11:54, Paul ***@***.***> schrieb:

Hi,

I'm trying to get ngx-quill working in my project for the first time, using
Angular 11. It all seems to look good, but binding using formControlName
seems to fail. The following (simplified) HTML code has replaced my
textarea code that works fine with the formControlName attribute :

  <form *ngIf="form" [formGroup]="form" (submit)="submit()">
                    <div formArrayName="templates">
                        <div class="card mt-3" *ngFor="let template of
templates.controls; let i=index;"
                            [formGroupName]="i">
                           ...
                            <div class="card-body">
                              ....
                                  <quill-editor #editor [modules]="modules"

(onEditorCreated)="editorCreated($event)" placeholder="Enter Text"

(onSelectionChanged)="onSelectionChanged($event)"

(onContentChanged)="onContentChanged($event)" name="body"

formControlName="body"></quill-editor>
                              ....
                              </div>
                        </div>
                 </div>
          </form>

The full error I get is:

core.js:6157 ERROR Error: No value accessor for form control with
path: 'templates -> 0 -> body'

I know that error from past experiences building my own controls. I noticed
the following fixed the issue with my controls, but is not implemented in
ngx-quill:

  constructor(private control: NgControl) {
    this.control.valueAccessor = this;
  }

Could you let me know what I'm doing wrong or perhaps my suggestion helps
to solve the problem? Let me know if you need some more info.

Cheers,
Paul

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/KillerCodeMonkey/ngx-quill/issues/1212>, or unsubscribe

**Comment by AppiePau** [comment]
Yes, I'm sure. The previous textarea that was in place, worked like a charm. Example:

```html
<textarea type="text" class="form-control" formControlName="body" style="height: 400px;"></textarea>
```
Works fine instead of the quill-editor component.

Please note that the quill control is used in the  context of a form and then in an array (`formArrayName="templates"`) and iterated using`*ngFor="let template of templates.controls; let i=index;" [formGroupName]="i"`. Hence the exact path is 'form -> templates -> 0 -> body' which is also shown in the error.

I'm not sure what causes this, but my best guess is that the valueAccessor is not set in the constructor of the editor component. I took a quick look in the editor component code and saw nothing like `this.control.valueAccessor = this;`. If you could tell me how to build the project like it is published on npm, I could do a pull request en tinker with it.

Cheers,
Paul

**Comment by KillerCodeMonkey** [comment]
i do not think that this is needed for a custom value accessor, because you declare your component as a value accessor and thats it.

i built a quick example and it is working as expected.
https://github.com/KillerCodeMonkey/ngx-quill-example/blob/master/src/app/reactive-forms/reactive-forms.component.ts#L28
https://github.com/KillerCodeMonkey/ngx-quill-example/blob/master/src/app/reactive-forms/reactive-forms.component.html#L38

It is already live: https://killercodemonkey.github.io/ngx-quill-example/

search for "complex".

**Comment by AppiePau** [comment]
Hi,

Thanks for working out a proof of concept so fast! I wanted to add some feedback on the issue for future references in this ticket. It took some time to figure out, but the issue has been resolved!

I was under the presumption that the issue was caused by lazy loading content and generating the form (FormGroup) dynamically. That turned out to be a false notion. Then some other issues surfaced in the project, mainly some npm dependency version issues that caused an error 'record.factory is not a function'. Not related, but really set me off on the wrong foot.

Well, long story short, I fixed the error by forcing npm to install the correct versions (angular cli and typescript). When I ran into "formControlName: No value accessor for form control with path" again, I tried adding the QuillModule to the lowest module (eg. TemplateModule) which fixed all issues. Turns out I was just chasing the wrong tail.

For the record, I have `QuillModule.forRoot(...)` still in my AppModule to config the editor.

Cheers!
